migrator/migrator.go

```diff
@@ -0,0 +1,97 @@
+package migrator
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/microstorage"
+)
+
+const (
+	repoSlug = "github.com/giantswarm/microstorage"
+
+	migrationKey = repoSlug + "/migration"
+	migrationVal = "v1"
+)
+
+type Config struct {
+	Logger micrologger.Logger
+}
+
+// DefaultConfig creates a new configuration with the default settings.
+func DefaultConfig() Config {
+	return Config{
+		Logger: nil, // Required.
+	}
+}
+
+type Migrator struct {
+	logger micrologger.Logger
+}
+
+func New(config Config) (*Migrator, error) {
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.Logger is empty")
+	}
+
+	m := &Migrator{
+		logger: config.Logger,
+	}
+
+	return m, nil
+}
+
+func (m *Migrator) Migrate(ctx context.Context, dst, src microstorage.Storage) error {
+	var err error
+
+	m.logger.Log("debug", "checking if migraiton is already done")
+	{
+		v, err := dst.Search(ctx, migrationKey)
+		if microstorage.IsNotFound(err) {
+			// This means migration hasn't been done yet.
+		} else if err != nil {
+			return microerror.Maskf(err, "dst storage: searching key=%s", migrationKey)
+		} else if v == migrationVal {
+			m.logger.Log("info", "migration already done")
+			return nil
+		}
+	}
```

Comment:
> Operators do not create additional state to ensure that resources do not have to be reconciled anymore.

Because they don't have to. The side effect is the created resource - this is the state. I can easily imagine an operator which has to create state in order to track the resources. This is what we wanted to do for consistent deletions. Even setting a finalizer means setting state on managed resource.

In this case, this is here to make the operation idempotent.

I want to call `Migrate` on the service start for the "migration period". If we don't have a flag that says the migration is done, we can override new values with the old from legacy storage, when the service restarts. Example:

1. Service pod starts and running app does the migration on boot.
2. The app running inside a pod does some updates to a key `k1`.
3. The pod dies.
4. The pod restarts and the app tires to migrate the data, but it sees that mark is set so does nothing and updated key `k1` isn't overridden.

This is the smartest solution I could think of. I'm open to suggestions.

Theres is a race possible when more than two replicas start, but this is rather safe, as during the migration only `Put` operation is invoked on the new storage, and probability of data modification is small during that time. The time window is between: one replica finished migration while other one is still doing it.